lib/collection/src/shards/local_shard/query.rs

```diff
@@ -32,15 +35,60 @@ impl LocalShard {
             .query_scroll_batch(request.scrolls, search_runtime_handle)
             .await?;
 
-        self.recurse_prefetch(
-            &request.merge_plan,
-            &core_results,
-            &scrolls,
-            search_runtime_handle,
-            timeout,
-            0, // initial depth
-        )
-        .await
+        let mut scored_points = self
+            .recurse_prefetch(
+                &request.merge_plan,
+                &core_results,
+                &scrolls,
+                search_runtime_handle,
+                timeout,
+                0, // initial depth
+            )
+            .await?;
+
+        // fetch payload and/or vector for scored points if necessary
+        if request.with_payload.is_required() || request.with_vector.is_enabled() {
+            // ids to retrieve
+            let mut point_ids = Vec::new();
+            for scored_point in scored_points.iter().flatten() {
+                point_ids.push(scored_point.id);
+            }
+            // remove duplicates
+            point_ids.sort_unstable();
+            point_ids.dedup();
```

Comment:
isn't it better to use a `HashSet` in this case?

Reply:
The number of points here is rather small, so I decided to go with the `vec` to avoid creating a lot of small `hashset` which consume more memory.

Reply:
If we do group-by in the same high-level way we are doing it right now, it will be normal to have a high limit. It will be `limit` * `group_size` in each iteration

Reply:
Alright, I switched to a HashSet :)